cdap-ui/app/hydrator/controllers/detail-ctrl.js

```diff
@@ -15,65 +15,67 @@
  */
 
 angular.module(PKG.name + '.feature.hydrator')
-  .controller('HydratorPlusPlusDetailCtrl', function(HydratorPlusPlusDetailRunsStore, rPipelineDetail, HydratorPlusPlusDetailActions, $scope, HydratorPlusPlusDetailNonRunsStore, HydratorPlusPlusDetailMetricsActions, $stateParams, PipelineAvailablePluginsActions) {
+  .controller('HydratorPlusPlusDetailCtrl', function(rPipelineDetail, $scope, HydratorPlusPlusDetailMetricsActions, $stateParams, PipelineAvailablePluginsActions, GLOBALS) {
     // FIXME: This should essentially be moved to a scaffolding service that will do stuff for a state/view
-    HydratorPlusPlusDetailRunsStore.init(rPipelineDetail);
-    HydratorPlusPlusDetailNonRunsStore.init(rPipelineDetail);
-
+    const actionCreator = window.CaskCommon.PipelineDetailActionCreator;
     this.pipelineType = rPipelineDetail.artifact.name;
+    let programType = this.pipelineType === GLOBALS.etlDataPipeline ? 'workflows' : 'spark';
+    let programId = this.pipelineType === GLOBALS.etlDataPipeline ? 'DataPipelineWorkflow' : 'DataStreamsSparkStreaming';
+
+    let currentRunId;
+    let pluginsFetched = false;
 
+    actionCreator.init(rPipelineDetail);
     let runid = $stateParams.runid;
     if (runid) {
-      HydratorPlusPlusDetailActions.setCurrentRun(runid);
+      actionCreator.setCurrentRun(runid);
     }
-    var params = HydratorPlusPlusDetailRunsStore.getParams();
-    params.scope = $scope;
-    var currentRunId;
-    this.PipelineAvailablePluginsActions = PipelineAvailablePluginsActions;
 
-    this.PipelineAvailablePluginsActions.fetchPluginsForDetails($stateParams.namespace, rPipelineDetail.config.stages);
+    actionCreator.getRuns({
+      namespace: $stateParams.namespace,
+      appId: rPipelineDetail.name,
+      workflowId: programId
+    });
 
-    HydratorPlusPlusDetailRunsStore.registerOnChangeListener(function () {
+    this.pipelineDetailStoreSubscription = window.CaskCommon.PipelineDetailStore.subscribe(() => {
+      let newPipelineDetailStoreState = window.CaskCommon.PipelineDetailStore.getState();
+
+      if (!pluginsFetched) {
+        PipelineAvailablePluginsActions.fetchPluginsForDetails($stateParams.namespace, newPipelineDetailStoreState.config.stages);
+        pluginsFetched = true;
+      }
```

Comment:
does this have to wait until the store subscription? Seems like you can get the list of stages from `rPipelineDetail`. Then you won't need this flag `pluginsFetched` anymore

Reply:
Originally `rPipelineDetail.config.stages` was available because of this line https://github.com/caskdata/cdap/blob/develop/cdap-ui/app/hydrator/services/detail/stores/pipeline-detail-nonruns-store.js#L305. Basically in that store, after we parsed `app.configuration` (`app` being `rPipelineDetail` in this case), we set `app.config` to be equal to the parsed value. However, this seemed dirty, so I have moved away from that in my new store.

So here I still need to wait for the store subscription, because I cannot get list of stages from `rPipelineDetail` anymore.